reconcile/utils/differ.py

```diff
@@ -0,0 +1,52 @@
+from collections.abc import (
+    Callable,
+    Mapping,
+)
+from dataclasses import dataclass
+from typing import (
+    Generic,
+    Optional,
+    TypeVar,
+)
+
+C = TypeVar("C")
+D = TypeVar("D")
+K = TypeVar("K")
+
+
+@dataclass(frozen=True, eq=True)
+class DiffResult(Generic[C, D, K]):
+    add: dict[K, D]
+    delete: dict[K, C]
+    change: dict[K, tuple[C, D]]
+
+
+def diff(
+    current: Mapping[K, C],
+    desired: Mapping[K, D],
```

Comment:
Why are you using `Mapping`s as parameter? Most of the time, we've different `list` to compare to.

Reply:
This is based on https://github.com/app-sre/qontract-reconcile/blob/8257677076a73eb79443559747edef376e872c93/reconcile/openshift_base.py#L585

which I worked on some parts before, where are we using `list` diff?

Reply:
@chassing How about 2 functions, one for diff dict, another for diff lists?

Reply:
Here are some examples:

* https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/ocm_oidc_idp.py#L48
* https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/aus/base.py#L108
* https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/openshift_users.py#L102
* https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/openshift_groups.py#L114

Reply:
> @chassing How about 2 functions, one for diff dict, another for diff lists?

Sounds good to me!

Reply:
Now there are 3 functions for different usages:
* `diff_mappings` for 2 dicts
* `diff_by_key` for 2 iterables, this is handy when no `__eq__` and `__hash__` overrides
* `diff_lists` for 2 lists, can be simple elements or with `__eq__` and `__hash__` overrides

Reply:
I think you should return a standard `DiffResult` class instead of differentiating between mappings and lists. I think the most common use-case will be something like the following. Imaging `Foo` is class from `fetch_desired_state`:

```python
@dataclass
class Foo:
    name: str
    value: int

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Foo):
            return False

        return self.name == other.name and self.value == other.value

    def __hash__(self) -> int:
        return hash(self.name)


a_foos = [Foo(name="foo", value=1), Foo(name="bar", value=2)]
b_foos = [Foo(name="foo", value=2), Foo(name="lala", value=2)]

print(
    diff(
        a_foos,
        b_foos,
        key=lambda x: x.name,
    )
)
# DiffResult(add={'lala': Foo(name='lala', value=2)}, delete={'bar': Foo(name='bar', value=2)}, change={'foo': Foo(name='foo', value=2)})

```

and 
```python
def diff(current, desired, key, equal: Callable = None) ...
    if not equal:
        equal = lambda x, y: x == y
    ...
```

And having just the desired item in `change`. What do you think?

Reply:
* How to use one `DiffResult` to support both dict and list? If diff 2 lists without key func like https://github.com/app-sre/qontract-reconcile/blob/8257677076a73eb79443559747edef376e872c93/reconcile/openshift_users.py#L131 

* I usually like to write immutable code, once a variable is assigned, never assign it to other value, reassign function parameter sometimes cause confusion, maybe OK in this case as it's very common in python to set default value this way, just a habit from eslint rule [no-param-reassign](https://eslint.org/docs/latest/rules/no-param-reassign)

* in some cases we want to zip `current` with `desired` to do a compare, for example https://github.com/app-sre/qontract-reconcile/blob/8257677076a73eb79443559747edef376e872c93/reconcile/ocm_oidc_idp.py#L141, it will be very handy to return a tuple in `change` than retrieve them later like current code using `O(N^2)` time complexity.

Reply:
and sometimes current and desired are not using the same type, that's why I split them and provide two key func as parameters, the signature is very similar to `Join` in [Linq](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.join).

Reply:
> How to use one DiffResult to support both dict and list? ...
I mean having the same attributes: add, delete, change. In your `DiffListResult` you have identical instead of change.

>  If diff 2 lists without key func like

`key=lamda x: x`?

> ... maybe OK in this case as it's very common in python to set default value this way ...

Yes, that's is very common pattern in python.

> ... it will be very handy to return a tuple in change than retrieve them ...

But this isn't very readable:

```python
changes = diff(
    a_foos,
    b_foos,
    key=lambda x: x.name,
)

delete_foos(changes.delete)
add_foos(changes.add)
update_foos(changes.change[1]) <- I don't like that
```

Reply:
I use `identical` instead of `change` for list because that's the result of intersect, name it `change` seems strange, they are `__eq__` but maybe changed... To keep structure consistent, I think rename it to `change` is OK. I'll see if I can use a single `DIffResult` class using generic now that they have same fields.

> key=lamda x: x

Good idea, then maybe we can always use the old `DiffResult` structure. 

Regarding `change`, how about use named tuple or another dataclass with `current` and `desired` as fields?

Reply:
> Regarding change, how about use named tuple or another dataclass with current and desired as fields?

Yes, sure, sounds good. 🙇